test/e2e/storage/subpath.go

```diff
@@ -310,6 +317,7 @@ var _ = utils.SIGDescribe("Subpath", func() {
 				pod.Spec.Volumes[0].VolumeSource = *roVol
 
 				// Read it from inside the subPath from container 0
+				pod.Spec.Containers[0].VolumeMounts[0].ReadOnly = true
```

Comment:
what happens when volumeMount doesn't specify readOnly but volumeSource does? from what I understand it will silently mount the subpath as readWrite? Not sure if that's behavior I would expect. Your comment was "...if the base volume is set to readOnly, then the user must also set all subpath mounts from it to readOnly too." Is there code that makes sure this is true?

Either way, would like to see a test case of this, "readOnly specified in volumeSource but not volumeMount" with the expected behavior there. I thought you already wrote one like that before but I can't find it right now for some reason.

Reply:
If volume source is readOnly, but volume mount of the subpath is not, then the subpath mount will fail because it will try to mount rw, but the underlying volume is ro.

I can add a new negative test case for this.  The original fix I made actually supported this scenario, and I added e2e tests for it, but that broke certain atomic volume configs and SElinux env.  So in this fix, I modified those original e2e tests to also set VolumeMount.ReadOnly to true if the volume source is also marked readonly.

Reply:
I see, LGTM after negative test added :)

Reply:
Ok I tried to add these new negative test cases, but actually the pods succeeded when I expected them to fail.  They are still read only in the container though.  Digging further into this, it looks like the extra "bind" argument that was added to the bindRemountOpts is causing the subpath to be mounted with the same read/write settings as the underlying mount.

Here's the gce pd mounts when I specified readOnly = true for the gce pd volumeSource and not the subpath.
```
/dev/sdb on /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/msau-disk type ext4 (ro,relatime,data=ordered)
/dev/sdb on /var/lib/kubelet/pods/cce2531a-6433-11e8-96df-42010a800002/volumes/kubernetes.io~gce-pd/vol1 type ext4 (ro,relatime,data=ordered)
/dev/sdb on /var/lib/kubelet/pods/cce2531a-6433-11e8-96df-42010a800002/volume-subpaths/vol1/subpath1/0 type ext4 (ro,relatime,data=ordered)
```

Here's configmap mounts when no read only is specified:
```
/dev/sda1 on /var/lib/kubelet type ext4 (rw,relatime,commit=30,data=ordered)
/dev/sda1 on /var/lib/kubelet/pods/7bb2dff0-6436-11e8-96df-42010a800002/volume-subpaths/vol1/subpath1/0 type ext4 (rw,relatime,commit=30,data=ordered)
```
Note that atomic volumes has special handling to make the runtime bind mount it as ro.

When I tried removing the "bind" argument from the remount, pod startup failed as expected:
```
May 30 05:49:54 e2e-test-msau-minion-group-hr81 kubelet[1937]: E0530 05:49:54.711195    1937 kubelet_pods.go:215] failed to prepare subPath for volumeMount "vol1" of container "subpath1": error mounting /var/lib/kubelet/pods/0fdbfd32-63cd-11e8-9a6b-42010a800002/volumes/kubernetes.io~gce-pd/vol1/dir1: mount failed: exit status 32
May 30 05:49:54 e2e-test-msau-minion-group-hr81 kubelet[1937]: Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/0fdbfd32-63cd-11e8-9a6b-42010a800002/volume-subpaths/vol1/subpath1/0 --scope -- mount -o remount /proc/1937/fd/36 /var/lib/kubelet/pods/0fdbfd32-63cd-11e8-9a6b-42010a800002/volume-subpaths/vol1/subpath1/0
```
And only when I also specify readOnly: true in the subpath volumeMount, that's when pod bringup succeeded.

Some more test cases to try 
* base volume has volumeMount readOnly, subpath does not
* subpath to non-existent file when volumeSource is readOnly

Reply:
I updated the tests to add:
* When volumeSource is marked readOnly, subpath doesn't need to be marked as readOnly for existing paths
* When volumeSource is marked readOnly, subpath should fail if path doesn't exist

Reply:
so the solution here is to change the behavior from what was originally stated to "when volume source is marked readOnly, but subpath is not, the pod should NOT fail and the subpath mounts will be readOnly". Should this be documented somewhere? It seems like reasonable default behavior to me so maybe it doesn't need to be called out specifically.

Reply:
Also I don't see the new tests, have they been added yet?

Reply:
The solution here is to not change the previous behavior.  If the volumeSource.readOnly flag is set, then the subpath will inherit that setting without needing the user to explicitly set it in volumeMount.  I've updated the PR description to describe the whole behavior a little better.

All the test changes are in the e2e file.